### ISSUE TITLE
[Feature] Option to automatically move to nearest mark during mark navigation

### ISSUE THREAD

**Issue by pidgeon777** [LeonHeidelbach/trailblazer.nvim#24]
### Did you check the docs?

- [X] I have read all the trailblazer.nvim docs

### Is your feature request related to a problem? Please describe.

Let's consider a stack with `global_fpath_line_sorted` mode active, and a buffer with some marks.

This could be the mark list:

```
1: file1.txt:10
2: file1.txt:20
3: file1.txt:30
4: file1.txt:40
```

Now I execute the `TrailBlazerPeekMoveNextDown` command, new situation:

```
1: file1.txt:10
2: file1.txt:20 <--
3: file1.txt:30 
4: file1.txt:40  
``` 

Again:

```
1: file1.txt:10
2: file1.txt:20
3: file1.txt:30 <--
4: file1.txt:40  
```

Now I decide to move the cursor at line `15`, and I execute the `TrailBlazerPeekMoveNextDown` or `TrailBlazerPeekMoveNextDown` command, obtaining:

```
1: file1.txt:10
2: file1.txt:20
3: file1.txt:30 <--
4: file1.txt:40
```

With the actual behaviour, the cursor goes back to the previously selected mark (`#3`), which is at line `30`.

### Describe the solution you'd like

The wished behaviour could be, in the case of `TrailBlazerPeekMoveNextDown` command:

```
1: file1.txt:10
2: file1.txt:20 <--
3: file1.txt:30
4: file1.txt:40
```

and, in the case of `TrailBlazerPeekMovePreviousUp` command:

```
1: file1.txt:10 <--
2: file1.txt:20
3: file1.txt:30
4: file1.txt:40
```

Thus, switching to the previous or next mark closest to the actual cursor position, according to the mark navigation command. The trail mark list would be updated accordingly.

Maybe it wouldn't make sense to enable this optional mode for the `chron` trail mark selection modes, but for the remaining, yes.

### Describe alternatives you've considered

N/A

**Comment by pidgeon777** [comment]
What if the `global_fpath_line_sorted` is active and the cursor is placed in a buffer with no marks?

Then, the path of the current buffer should be kept into account, and with the `TrailBlazerPeekMovePreviousUp` command, the cursor should be placed to the latest mark of the nearest buffer sorted by path.

Similar behaviour for `TrailBlazerPeekMoveNextDown` command, but in this case, the cursor should be placed to the earliest mark of the nearest buffer sorted by path.

Example of mark list:

```
/path1/file.txt:10
/path3/file.txt:40
```
If the cursor is placed in a buffer with path `/path2/file.txt`, then:

- `TrailBlazerPeekMovePreviousUp` should place the cursor on the mark at `path1/file.txt:10`.
- `TrailBlazerPeekMoveNextDown` should place the cursor on the mark at `path3/file.txt:40`.

What if a buffer has no path? Then, `TrailBlazerPeekMoveNextDown` and `TrailBlazerPeekMovePreviousUp` should behave as usual, selecting the next/previous mark in the mark list.

All of this (even the proposal of my previous post) by letting the user set a single variable to choose the behaviour he wants (for example, `auto_nearest` set to `true` or `false`).

**Comment by LeonHeidelbach** [comment]
@pidgeon777 I actually implemented this feature a couple of days ago 8d4eae117c3e7eea7ca2327884cb2c86abbce851. Just look at the updated [mappings table](https://github.com/LeonHeidelbach/trailblazer.nvim#%EF%B8%8F-configuration). In the motions section you will find the new key binding `move_to_nearest` which does exactly that. You can also use the new user command `:TrailBlazerMoveToNearest`. This new function calculates the minimum Manhattan Distance from your cursor position to all marks within the current buffer and moves the cursor to it if one exists. Otherwise nothing happens. I thought about moving to the trail mark cursor in a different file (just as peek moving does) if there is no closest mark and implemented this first, but this is in my opinion not the expected behavior since you could just use the move commands to achieve this.

**Comment by pidgeon777** [comment]
The problem with `TrailBlazerMoveToNearest` in my opinion is the following. Starting situation, with current cursor position circled in red:

![image](https://user-images.githubusercontent.com/13739892/220593214-d4f6ed0b-1547-4bd1-80d3-150988d03758.png)

Let's suppose I want to move here:

![image](https://user-images.githubusercontent.com/13739892/220592174-d2540644-2604-45b9-b361-bcd0df1257de.png)

Then by first executing the `TrailBlazerMoveToNearest` command, the cursor would be placed here first, because the nearest:

![image](https://user-images.githubusercontent.com/13739892/220592436-99ef3b6d-866e-4e38-9900-5f0c5c7f1f12.png)

But logic would impose that if I want to navigate to the previous mark starting from the cursor position, it wouldn't make sense to go to the next one first only because closer, then call again the `TrailBlazerPeekMovePreviousUp` command to go to the wished location.

So in my opinion maybe a different command could be added, or again an option which modifies the behaviour of the `TrailBlazerPeekMoveNextDown` and `TrailBlazerPeekMovePreviousUp` commands only.

Also, in my opinion, keeping projects into account, this modified behaviour for the two commands would help when the cursor is placed for example on a project buffer with no marks. Logic would imply that executing the `TrailBlazerPeekMoveNextDown` and `TrailBlazerPeekMovePreviousUp` commands, then the nearest path sorted buffer with marks would be selected, which would greatly help with directly switching to the closed buffer of interest belonging to the project, given that it would share the base path of the starting buffer.

Still, if the current buffer has no path, the two mark navigation commands would work as always.

So, in my opinion, the `TrailBlazerMoveToNearest` is good and should be left as is for what it does: moving to the closest mark available, without switching to any different buffer.

But, the setting I propose, is to modify the behaviour of the `TrailBlazerPeekMoveNextDown` and `TrailBlazerPeekMoveNextDown` commands only.

**Comment by LeonHeidelbach** [comment]
If I understood you correctly, the proposed behavior of `:TrailBlazerPeekMovePreviousUp` and `:TrailBlazerPeekMoveNextDown` would then no longer allow users to go back to the current trail mark cursor position within the stack if they wanted to. This would cause a major inconsistency in the plugin's general behavior and render the trail mark cursor virtually useless. I don't think adding an option to modify the peek motions is the way to go here. The way I could see this potentially work is by allowing the user to specify an additional argument to the `:TrailBlazerMoveToNearest` user command that defines the movement direction if no nearest mark was found within the current buffer (i.e. `fpath_up`, `fpath_down`,`buf_up`,`buf_down`) .

**Comment by pidgeon777** [comment]
The way I could see this potentially work is by allowing the user to specify an additional argument to the :TrailBlazerMoveToNearest user command that defines the general movement direction (i.e. fpath_up, fpath_down,buf_up,buf_down).

Yes, could be a good idea 👍. But how those options would behave in normal (buffer with marks) and "special" situations, for example, the ones where the cursor is placed on a buffer with no marks, or on a buffer with no path?

Also, it could be possible to implement that in the API?

**Comment by LeonHeidelbach** [comment]
> Yes, could be a good idea +1. But how those options would behave in normal (buffer with marks) and "special" situations, for example, the ones where the cursor is placed on a buffer with no marks, or on a buffer with no path?

They would behave just as the additional parameter would suggest, but within the current buffer context. This means that `fpath_up` would just become `up` and `fpath_down` just be `down`, so the nearest mark mark would only be searched for in the specified direction. If the buffer has no marks, the directive specifier `fpath` or `buf` would be used to determine the next move action. If the buffer has no name I think it would be best to just do nothing.
 
> Also, it could be possible to implement that in the API?

Everything that a usercommand can do is automatically available over the API, since I just pass the specified parameters when using a user command to the respective API functions.

**Comment by pidgeon777** [comment]
Sounds good to me 👍.

**Comment by LeonHeidelbach** [comment]
@pidgeon777 This has been implemented bb96107bd5efca522a8b7b4a4257715d627c595b. I left out the `buf_up` and `buf_down` directives in my implementation since this did not seem very useful. You can now use the API function like this to enable the proposed behavior:

```lua
 -- pass nil or 0 as the first argument to use the current buffer as the base for the motion
require("trailblazer").move_to_nearest(nil, "fpath_down")
```

You can find all available directives for the `move_to_nearest` function [here](https://github.com/LeonHeidelbach/trailblazer.nvim#trailblazermovetonearest-motion-directives).

**Comment by pidgeon777** [comment]
I think we're almost there, please notice the following video:

![wAi81IB8ao](https://user-images.githubusercontent.com/13739892/220998011-93002e28-87fe-4dac-865c-1c180ac9ff55.gif)

What I'm doing is continuously executing the `require("trailblazer").move_to_nearest(nil, "fpath_up")` command to move up in the mark list, then `require("trailblazer").move_to_nearest(nil, "fpath_down")` to move down.

When I reach the mark list limits, something strange happens.

It is not possible to reach the `.txt` buffer, even if it contains mark, executing `require("trailblazer").move_to_nearest(nil, "fpath_up")` just cycle between the `lua` files.

If I reach the last mark list item at its bottom, executing `require("trailblazer").move_to_nearest(nil, "fpath_down")` leads the mark pointer to move to the penultimate mark, then down to the last, then up to the penultimate, then back to the last.

**Comment by pidgeon777** [comment]
What I would expect is: if I reached the last item of the mark list, then `require("trailblazer").move_to_nearest(nil, "fpath_down")` shouldn't do anything. Also, I expected to reach the top marks in the list with `require("trailblazer").move_to_nearest(nil, "fpath_up")`, but it wasn't possible.

I think this "bounce" in the bottom of the list, could affect the top of the list, though: once the mark at the top of the list is reached, `require("trailblazer").move_to_nearest(nil, "fpath_up")` shouldn't do anything.

**Comment by pidgeon777** [comment]
After saving the stack and re-opening it, now it is possible to access even the marks of the `.txt` file, but a bounce occurs between the first and the third mark:

![QdfqBRKOx7](https://user-images.githubusercontent.com/13739892/221003767-192a4aae-83ae-49cd-9e0a-2136f0840f42.gif)

Maybe we should also pay attention to possible issues associated with marks placed in the first columns.

**Comment by LeonHeidelbach** [comment]
Not sure what causes this exactly, I could reproduce this issue in some projects but not in others. The jump happens whenever the previous buffer could not be found and as a fallback I jump back to the first entry of the file list. I will have to investigate what causes this, but I don't know if I'll get to it today.

**Comment by LeonHeidelbach** [comment]
@pidgeon777 I think I potentially found out what the problem was, however it could also be some Lua internal weirdness with list indices that I have encountered before, that I now potentially fixed by not cloning a table. What happened was that if the first file in the list (lua index 1) was reached the fallback would reset the previous file to 1 again if there was no previous file. Moving further up would then trigger the upward motion that sets the cursor to the last trail mark of the "previous" file which would be the current one. I fixed this by checking if the previous file is the same as the current file which should never happen. It is however very strange that it happened in some projects but not in others, but these are things that can happen in lua land i guess :D. Let me know if the latest commit fixes the problem for you bd7653a431bc78cffb8543aba5237662eb36af36.

**Comment by pidgeon777** [comment]
Bounce at the bottom and top mak is fixed, now a problem arises if I place multiple marks on the same line and navigate with `require("trailblazer").move_to_nearest(nil, "fpath_up")` or `down`, in that case only 1 mark get "caught" for each line. See video:

![lkIWWbaWMg](https://user-images.githubusercontent.com/13739892/221130203-1145f95b-8604-4214-9618-8b090ddd2768.gif)

**Comment by pidgeon777** [comment]
Also, I suggest checking if the navigation still works even if placing a mark on column `0` for a given line, because I encountered a strange navigation behaviour during one of my tests if I recall correctly, which involved placing marks at column `0`.

**Comment by pidgeon777** [comment]
After doing some tests, I think something is broken. I added/removed some marks, and as result, I couldn't navigate to all of them.

I wonder if this could be related to some issue when managing relative/absolute paths, or when populating the mark list.

I'm navigating using:

```lua
require("trailblazer").move_to_nearest(nil, "fpath_up")
require("trailblazer").move_to_nearest(nil, "fpath_down")
```

and the mark mode is set to `global_fpath_line_sorted`.

**Comment by LeonHeidelbach** [comment]
> Bounce at the bottom and top mak is fixed, now a problem arises if I place multiple marks on the same line and navigate with `require("trailblazer").move_to_nearest(nil, "fpath_up")` or `down`, in that case only 1 mark get "caught" for each line.

This is actually the intended behavior. The `move_to_nearest` command combined with `fpath_up/down` currently only searches for the nearest marks in the defined directions but not on the same line.
 

> Also, I suggest checking if the navigation still works even if placing a mark on column `0` for a given line, because I encountered a strange navigation behaviour during one of my tests if I recall correctly, which involved placing marks at column `0`.

I placed multiple marks on column 0 and navigated using the `move_to_nearest` command and everything seemed to work correctly. What were the specific circumstances in which you encountered the issue?



> After doing some tests, I think something is broken. I added/removed some marks, and as result, I couldn't navigate to all of them.
> 
> I wonder if this could be related to some issue when managing relative/absolute paths, or when populating the mark list.
> 
> I'm navigating using:
> 
> ```lua
> require("trailblazer").move_to_nearest(nil, "fpath_up")
> require("trailblazer").move_to_nearest(nil, "fpath_down")
> ```
> 
> and the mark mode is set to `global_fpath_line_sorted`.

I did the exact same thing, but could not reproduce the issue. Can you give me more info on when this happens for you exactly? This should not be a problem with different file path formats since they are all normalized using the `vim.api.expand` function and the file list is currently created anew on every file jump.

**Comment by LeonHeidelbach** [comment]
@pidgeon777 I have now added functionality that respects closest marks on the same line when using `move_to_nearst` + directives d430615405e403bc9f71103d0ab3872c525fbba5.

This video shows the correct jump movements using `peek_moves` first (`<A-J>` & `<A-K>` keys) and then the same correct moves using `move_to_nearest` jumps (`<A-z>` & `<A-u>` keys). Notice that the cursor traverses all marks when using `peek_moves` in the beginning but not when using `move_to_nearest` jumps since the distance from the current cursor position to the nearest mark does not always represent the actual mark order, thus making some marks unreachable in some cases.

https://user-images.githubusercontent.com/55194243/221166884-3e9e7584-ad5a-4758-84ab-f25760bee626.mp4

**Comment by LeonHeidelbach** [comment]
@pidgeon777 Btw. what tool are you using to record your GIFs. I use OBS and convert the recording to GIF throuh ffmpeg but then my recordings get some kind of a weird grid pattern glitch as you can see on my demo GIF in the README (Chrome tries to smooth this out though it's only really visible on other browsers). Yours don't seem to have same issue.

**Comment by pidgeon777** [comment]
The issue (not really an issue) with the `TrailBlazerPeekMovePreviousUp` and `TrailBlazerPeekMoveNextDown` commands, is that those commands don't keep into account the actual position of the cursor when called, but they refer to the actual "pointer" in the mark list, so executing them lead to switching to the previous or next item of the list, no matter where the cursor is.

`TrailBlazerMoveToNearest` is good, but it would be great to add the `linear` option to distinguish it from the `manhattan` radial analyses. `linear` would be much simpler than `manhattan`, because it would simply involve scanning for the next/previous mark starting from the cursor in a linear way as if the buffer was made of a single line.

**Comment by LeonHeidelbach** [comment]
@pidgeon777 I added an additional parameter to the `move_to_nearest` function called `dist_type`. This allows users to choose between the type of distance calculation. The default is `man_dist` but it can be changed to `lin_char_dist` to do what you want.

```lua
-- pass nil or 0 as the first argument to use the current buffer as the base for the motion
-- the third parameter can be omitted if you want to use the Manhattan Distance
require("trailblazer").move_to_nearest(nil, "fpath_up", "lin_char_dist")
-- and
require("trailblazer").move_to_nearest(nil, "fpath_down", "lin_char_dist")
```

**Comment by LeonHeidelbach** [comment]
> > @pidgeon777 Btw. what tool are you using to record your GIFs.
> 
> This one, very good:
> 
> https://getsharex.com/

@pidgeon777 Sadly, there doesn't seem to be a Linux version of this.

**Comment by pidgeon777** [comment]
> @pidgeon777 I added an additional parameter to the `move_to_nearest` function called `dist_type`. This allows users to choose between the type of distance calculation. The default is `man_dist` but it can be changed to `lin_char_dist` to do what you want [271ffcc](https://github.com/LeonHeidelbach/trailblazer.nvim/commit/271ffccc648cb6968a727d7e4c65e926808412ff).
> 
> ```lua
> -- pass nil or 0 as the first argument to use the current buffer as the base for the motion
> -- the third parameter can be omitted if you want to use the Manhattan Distance
> require("trailblazer").move_to_nearest(nil, "fpath_up", "lin_char_dist")
> -- and
> require("trailblazer").move_to_nearest(nil, "fpath_down", "lin_char_dist")
> ```

So far it seems to work great, and exactly what I wanted to achieve.

I'll do more tests and report here any found issues.

Awesome work as always 🔝.

**Comment by LeonHeidelbach** [comment]
@pidgeon777 @FelixKratz I introduced a few new configuration options that allow users to directly combine the `move_to_neareset` functionality with both `peek_move` motions if they so choose fb799e8afeefff06e2f15cdd38243860b2cc5332. This however, effectively "disables" the current trail mark cursor within the stack, since after enabling this feature you will always move to the nearest trail mark following the set `motion directive` and `dist_type`. Technically you can still move to the last trail mark cursor position by placing the cursor on any other trail mark that is not also the current trail mark cursor but you would have to do this intentionally. I left this behavior in for now, but let me know if this is either a good idea to have or should also be changed so that the trail mark cursor will be set to the mark under the cursor when a `peek_motion` is invoked. These are the default values with the feature enabled.

```lua
move_to_nearest_before_peek = true, -- default is false
move_to_nearest_before_peek_motion_directive_up = "fpath_up", -- "up", "fpath_up" -> For move info see section "TrailBlazerMoveToNearest Motion Directives"
move_to_nearest_before_peek_motion_directive_down = "fpath_down", -- "down", "fpath_down" -> For move info see section "TrailBlazerMoveToNearest Motion Directives"
move_to_nearest_before_peek_dist_type = "lin_char_dist", -- "man_dist", "lin_char_dist" -> Manhattan Distance or Linear Character Distance
```

If you don't want to change the default `motion directives` or `dist_type` you will only have to add `move_to_nearest_before_peek = true,` to your configuration table. Using this instead of completely switching to `move_to_nearest` motions for navigation allows you to still follow the mark order of other selection modes while also allowing for direct jumps to nearest marks when the cursor is currently not located on a trail mark.

**Comment by LeonHeidelbach** [comment]
> Technically you can still move to the last trail mark cursor position by placing the cursor on any other trail mark that is not also the current trail mark cursor but you would have to do this intentionally. I left this behavior in for now, but let me know if this is either a good idea to have or should also be changed so that the trail mark cursor will be set to the mark under the cursor when a `peek_move` motion is invoked.

I found this to be rather distracting and implemented the proposed alternative behavior 3ff4453d2579ca825acc97d05d24ecadbba9a3d5. The cursor now properly moves in the specified direction. This also means that the current trail mark cursor can now never be reached if it's not also the nearest mark to the cursor in the specified direction.

**Comment by LeonHeidelbach** [comment]
To separately enable navigating to the current trail mark cursor I have implemented a new user command + API function 0e2591125e202cf0a3dd9239a4924bf8447291b6. There is no default key map for this but you can very easily add one to your configuration as thescribed in the [user commands](https://github.com/LeonHeidelbach/trailblazer.nvim#-user-commands) section.

**Comment by pidgeon777** [comment]
Thanks for the updates! I'm a bit lost with those new settings, I would like to ask what kind of commands or API combinations I should execute for behaviour like this:

![image](https://user-images.githubusercontent.com/13739892/222108294-c1eceb4e-4201-4a36-92a1-7b8196594417.png)

- I place a cursor on a position with no mark (`green` circle).
- I execute `CMD_UP`.
- Cursor goes to the first mark available in the buffer `linearly` starting with the up direction (not with the `manhattan` method).
- I execute `CMD_UP` again (same command as before).
- If the actual buffer contains no marks, then the cursor is placed on the last mark of the first buffer containing marks encountered going up with the path.

This means that in this configuration, the mark quick fix list has to be like this:

```
path0/pos0 <- Final cursor position
path0/pos1
path1/pos0
path1/pos1
path2/pos0
<- Cursor placed here at first
path2/pos1
```

Where `path` is the buffer path, and `pos` is the mark position in that buffer.

In short:

- marks should be sorted first by buffer path, then by their position inside the buffer, and this order should be the same displayed in the mark quick fix list.
- all of the marks, even at column `0` or at the same line, have to be reached with the `CMD_UP` command, in `linear` mode.
- adding a mark, should accordingly update the mark quick fix list so that the new mark item is placed there keeping into account the actual buffer path and new mark position.

My question is, how to achieve this behaviour?

**Comment by LeonHeidelbach** [comment]
@pidgeon777 The described behavior should now be achievable by simply adding `trail_options.move_to_nearest_before_peek = true` to your configuration table.

**Comment by pidgeon777** [comment]
Works great, thanks for the tip 👍.
